File: src/backend-rust/src/routes/builds.rs
use std::collections::{HashMap, HashSet};

use axum::{
    Json, Router,
    extract::{Extension, Path, Query, State},
    http::{HeaderMap, StatusCode},
    response::Response,
    routing::{get, post},
};
use paladinscat_core::database::Database;
use serde_json::{Value, json};

use crate::{error::ApiError, request::RequestId};

use super::identity::{as_i64, json_response, parse_id, require_session, simple_error};

#[derive(Clone)]
struct BuildsState {
    database: Database,
}

pub fn router(database: Database) -> Router {
    Router::new()
        .route("/builds", get(list).post(create))
        .route("/builds/", get(list).post(create))
        .route("/builds/{id}", get(detail))
        .route("/builds/{id}/like", post(like))
        .with_state(BuildsState { database })
}

fn select_sql() -> &'static str {
    "SELECT b.id,b.user_id,b.champion_id,COALESCE(c.name,'Champion '||b.champion_id::text) AS champion_name, \
       b.name,b.items,b.cards,b.actives,b.talents,b.notes,b.visibility,b.likes,b.view_count,b.created_at,u.username \
     FROM builds b JOIN users u ON u.id=b.user_id LEFT JOIN champions c ON c.id=b.champion_id"
}

async fn select_build(
    state: &BuildsState,
    id: i64,
    request_id: &RequestId,
) -> Result<Option<Value>, ApiError> {
    state
        .database
        .one_json(&format!("{} WHERE b.id=$1", select_sql()), &[&id])
        .await
        .map_err(|error| ApiError::database(error, request_id))
}

fn int_array(value: Option<&Value>, maximum: usize) -> Option<Vec<i32>> {
    let Some(value) = value else {
        return Some(Vec::new());
    };
    let rows = value.as_array()?;
    if rows.len() > maximum {
        return None;
    }
    let mut seen = HashSet::new();
    let mut output = Vec::new();
    for row in rows {
        let id = as_i64(Some(row)).and_then(|value| i32::try_from(value).ok())?;
        if id <= 0 {
            return None;
        }
        if seen.insert(id) {
            output.push(id);
        }
    }
    Some(output)
}

fn build_cards(value: Option<&Value>) -> Option<Value> {
    let Some(value) = value else {
        return Some(json!([]));
    };
    let rows = value.as_array()?;
    if rows.len() > 5 {
        return None;
    }
    let mut seen = HashSet::new();
    let mut cards = Vec::new();
    for row in rows {
        let card_id = as_i64(row.get("card_id").or_else(|| row.get("cardId")))
            .and_then(|value| i32::try_from(value).ok())?;
        let level = as_i64(row.get("level").or_else(|| row.get("card_level")))
            .and_then(|value| i32::try_from(value).ok())?;
        if card_id <= 0 || !(1..=5).contains(&level) || !seen.insert(card_id) {
            return None;
        }
        cards.push(json!({"card_id":card_id,"level":level}));
    }
    Some(Value::Array(cards))
}

async fn list(
    State(state): State<BuildsState>,
    Extension(request_id): Extension<RequestId>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Response, ApiError> {
    let champion_id = query
        .get("championId")
        .and_then(|value| parse_id(value))
        .and_then(|value| i32::try_from(value).ok());
    let rows = if let Some(champion_id) = champion_id {
        state
            .database
            .query_json(
                &format!(
                    "{} WHERE b.champion_id=$1 AND b.visibility='public' ORDER BY b.likes DESC",
                    select_sql()
                ),
                &[&champion_id],
            )
            .await
    } else {
        state
            .database
            .query_json(
                &format!(
                    "{} WHERE b.visibility='public' ORDER BY b.likes DESC",
                    select_sql()
                ),
                &[],
            )
            .await
    }
    .map_err(|error| ApiError::database(error, &request_id))?;
    Ok(json_response(StatusCode::OK, Value::Array(rows)))
}

async fn create(
    State(state): State<BuildsState>,
    Extension(request_id): Extension<RequestId>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> Result<Response, ApiError> {
    let session = match require_session(&state.database, &headers, &request_id).await {
        Ok(session) => session,
        Err(response) => return Ok(response),
    };
    let champion_id = as_i64(body.get("champion_id"))
        .and_then(|value| i32::try_from(value).ok())
        .filter(|value| *value > 0);
    let name = body
        .get("name")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .trim()
        .to_owned();
    let items = int_array(body.get("items"), 4);
    let cards = build_cards(body.get("cards"));
    let actives = int_array(body.get("actives"), 4);
    let talents = int_array(body.get("talents"), 1);
    let visibility = if body.get("visibility").and_then(Value::as_str) == Some("private") {
        "private"
    } else {
        "public"
    };
    let Some(champion_id) = champion_id else {
        return Ok(simple_error(StatusCode::BAD_REQUEST, "Invalid champion_id"));
    };
    if name.is_empty() {
        return Ok(simple_error(
            StatusCode::BAD_REQUEST,
            "Build name is required",
        ));
    }
    let Some(items) = items else {
        return Ok(simple_error(
            StatusCode::BAD_REQUEST,
            "Builds can include up to 4 valid item IDs",
        ));
    };
    let Some(cards) = cards else {
        return Ok(simple_error(
            StatusCode::BAD_REQUEST,
            "Builds can include up to 5 cards with levels from 1 to 5",
        ));
    };
    let Some(actives) = actives else {
        return Ok(simple_error(
            StatusCode::BAD_REQUEST,
            "Invalid legacy active item IDs",
        ));
    };
    let Some(talents) = talents else {
        return Ok(simple_error(
            StatusCode::BAD_REQUEST,
            "Builds can include only 1 valid talent ID",
        ));
    };
    let notes = body
        .get("notes")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty());
    let inserted = state
        .database
        .one_json(
            "INSERT INTO builds(user_id,champion_id,name,items,cards,actives,talents,notes,visibility) \
             VALUES($1,$2,$3,$4,$5::jsonb,$6,$7,$8,$9) RETURNING id",
            &[
                &session.user_id,
                &champion_id,
                &name,
                &items,
                &cards,
                &actives,
                &talents,
                &notes,
                &visibility,
            ],
        )
        .await
        .map_err(|error| ApiError::database(error, &request_id))?
        .ok_or_else(|| ApiError::internal(&request_id))?;
    let id = as_i64(inserted.get("id")).ok_or_else(|| ApiError::internal(&request_id))?;
    let build = select_build(&state, id, &request_id)
        .await?
        .ok_or_else(|| ApiError::internal(&request_id))?;
    Ok(json_response(StatusCode::OK, build))
}

async fn detail(
    State(state): State<BuildsState>,
    Extension(request_id): Extension<RequestId>,
    Path(id): Path<String>,
) -> Result<Response, ApiError> {
    let Some(id) = parse_id(&id) else {
        return Ok(simple_error(StatusCode::NOT_FOUND, "Build not found"));
    };
    let Some(build) = select_build(&state, id, &request_id).await? else {
        return Ok(simple_error(StatusCode::NOT_FOUND, "Build not found"));
    };
    state
        .database
        .query_json(
            "UPDATE builds SET view_count=view_count+1 WHERE id=$1 RETURNING id",
            &[&id],
        )
        .await
        .map_err(|error| ApiError::database(error, &request_id))?;
    Ok(json_response(StatusCode::OK, build))
}

async fn like(
    State(state): State<BuildsState>,
    Extension(request_id): Extension<RequestId>,
    Path(id): Path<String>,
    headers: HeaderMap,
) -> Result<Response, ApiError> {
    let Some(id) = parse_id(&id) else {
        return Ok(simple_error(StatusCode::NOT_FOUND, "Build not found"));
    };
    let session = match require_session(&state.database, &headers, &request_id).await {
        Ok(session) => session,
        Err(response) => return Ok(response),
    };
    let mut client = state
        .database
        .connection()
        .await
        .map_err(|error| ApiError::database(error, &request_id))?;
    let transaction = client
        .transaction()
        .await
        .map_err(|error| ApiError::database(error.into(), &request_id))?;
    let existing = transaction
        .query_opt(
            "SELECT 1 FROM user_build_likes WHERE user_id=$1 AND build_id=$2 FOR UPDATE",
            &[&session.user_id, &id],
        )
        .await
        .map_err(|error| ApiError::database(error.into(), &request_id))?
        .is_some();
    let row = if existing {
        transaction
            .execute(
                "DELETE FROM user_build_likes WHERE user_id=$1 AND build_id=$2",
                &[&session.user_id, &id],
            )
            .await
            .map_err(|error| ApiError::database(error.into(), &request_id))?;
        transaction
            .query_opt(
                "UPDATE builds SET likes=GREATEST(likes-1,0) WHERE id=$1 RETURNING likes",
                &[&id],
            )
            .await
            .map_err(|error| ApiError::database(error.into(), &request_id))?
    } else {
        transaction
            .execute(
                "INSERT INTO user_build_likes(user_id,build_id) VALUES($1,$2)",
                &[&session.user_id, &id],
            )
            .await
            .map_err(|error| ApiError::database(error.into(), &request_id))?;
        transaction
            .query_opt(
                "UPDATE builds SET likes=likes+1 WHERE id=$1 RETURNING likes",
                &[&id],
            )
            .await
            .map_err(|error| ApiError::database(error.into(), &request_id))?
    };
    transaction
        .commit()
        .await
        .map_err(|error| ApiError::database(error.into(), &request_id))?;
    let likes = row
        .as_ref()
        .map(|row| row.get::<_, i32>("likes"))
        .unwrap_or_default();
    Ok(json_response(
        StatusCode::OK,
        json!({"liked":!existing,"likes":likes}),
    ))
}

File: legacy/src-backend/routes/builds.ts
import { FastifyInstance } from 'fastify';
import { query, one } from '../config/db';
import crypto from 'crypto';

function hashSessionToken(token: string): string {
  return crypto.createHash('sha256').update(token).digest('hex');
}

function normalizeIntArray(input: unknown, maxLength: number): number[] | null {
  if (!Array.isArray(input)) return [];
  if (input.length > maxLength) return null;
  const values = input.map((value) => Number(value));
  if (values.some((value) => !Number.isInteger(value) || value <= 0)) return null;
  return Array.from(new Set(values));
}

function normalizeBuildCards(input: unknown): Array<{ card_id: number; level: number }> | null {
  if (!Array.isArray(input)) return [];
  if (input.length > 5) return null;

  const seen = new Set<number>();
  const cards: Array<{ card_id: number; level: number }> = [];
  for (const value of input) {
    if (!value || typeof value !== 'object') return null;
    const row = value as { card_id?: unknown; cardId?: unknown; level?: unknown; card_level?: unknown };
    const cardId = Number(row.card_id ?? row.cardId);
    const level = Number(row.level ?? row.card_level);
    if (!Number.isInteger(cardId) || cardId <= 0 || !Number.isInteger(level) || level < 1 || level > 5) return null;
    if (seen.has(cardId)) return null;
    seen.add(cardId);
    cards.push({ card_id: cardId, level });
  }
  return cards;
}

async function selectBuild(buildId: number) {
  return one(
    `SELECT b.id, b.user_id, b.champion_id, COALESCE(c.name, 'Champion ' || b.champion_id::TEXT) AS champion_name,
        b.name, b.items, b.cards, b.actives, b.talents, b.notes, b.visibility, b.likes, b.view_count, b.created_at, u.username
       FROM builds b
       JOIN users u ON u.id = b.user_id
       LEFT JOIN champions c ON c.id = b.champion_id
       WHERE b.id = $1`,
    [buildId]
  );
}

export default async function buildsRoutes(fastify: FastifyInstance) {
  fastify.get('/', async (req: any) => {
    const championId = req.query.championId ? parseInt(req.query.championId as string, 10) : undefined;
    const q = championId ? 'WHERE b.champion_id = $1 AND b.visibility = \'public\'' : 'WHERE b.visibility = \'public\'';
    const params = championId ? [championId] : [];
    return query(
      `SELECT b.id, b.user_id, b.champion_id, COALESCE(c.name, 'Champion ' || b.champion_id::TEXT) AS champion_name,
        b.name, b.items, b.cards, b.actives, b.talents, b.notes, b.visibility, b.likes, b.view_count, b.created_at, u.username
       FROM builds b
       JOIN users u ON u.id = b.user_id
       LEFT JOIN champions c ON c.id = b.champion_id
       ${q}
       ORDER BY b.likes DESC`,
      params
    );
  });

  fastify.post('/', async (req: any, reply: any) => {
    const token = req.headers.authorization?.replace('Bearer ', '');
    const session = token
      ? await one('SELECT user_id FROM sessions WHERE token = $1 AND expires_at > now()', [hashSessionToken(token)])
      : null;
    if (!session) return reply.status(401).send({ error: 'Not authenticated' });

    const { champion_id, name, items, cards, actives, talents, notes, visibility } = req.body ?? {};
    const championId = Number(champion_id);
    const buildName = String(name ?? '').trim();
    const normalizedItems = normalizeIntArray(items, 4);
    const normalizedCards = normalizeBuildCards(cards);
    const normalizedActives = normalizeIntArray(actives, 4);
    const normalizedTalents = normalizeIntArray(talents, 1);
    const normalizedVisibility = visibility === 'private' ? 'private' : 'public';

    if (!Number.isInteger(championId) || championId <= 0) {
      return reply.status(400).send({ error: 'Invalid champion_id' });
    }
    if (!buildName) {
      return reply.status(400).send({ error: 'Build name is required' });
    }
    if (!normalizedItems) {
      return reply.status(400).send({ error: 'Builds can include up to 4 valid item IDs' });
    }
    if (!normalizedCards) {
      return reply.status(400).send({ error: 'Builds can include up to 5 cards with levels from 1 to 5' });
    }
    if (!normalizedActives) {
      return reply.status(400).send({ error: 'Invalid legacy active item IDs' });
    }
    if (!normalizedTalents) {
      return reply.status(400).send({ error: 'Builds can include only 1 valid talent ID' });
    }

    // The UI enforces the full Paladins deck rule (5 cards totaling 15 points).
    // The API keeps validation slightly looser so older or imported community
    // builds can still be represented, but it always stores card levels in a
    // lossless JSONB shape instead of overloading the legacy `actives` array.
    const inserted = await one(
      `INSERT INTO builds (user_id, champion_id, name, items, cards, actives, talents, notes, visibility)
       VALUES ($1, $2, $3, $4, $5::jsonb, $6, $7, $8, $9)
       RETURNING id`,
      [
        session.user_id,
        championId,
        buildName,
        normalizedItems,
        JSON.stringify(normalizedCards),
        normalizedActives,
        normalizedTalents,
        typeof notes === 'string' && notes.trim() ? notes.trim() : null,
        normalizedVisibility,
      ]
    );

    return selectBuild(Number(inserted.id));
  });

  fastify.get('/:id', async (req, reply) => {
    const id = parseInt((req.params as any).id, 10);
    const build = await selectBuild(id);
    if (!build) return reply.status(404).send({ error: 'Build not found' });
    await one('UPDATE builds SET view_count = view_count + 1 WHERE id = $1', [id]);
    return build;
  });

  fastify.post('/:id/like', async (req: any, reply: any) => {
    const id = parseInt(req.params.id as string, 10);
    const token = req.headers.authorization?.replace('Bearer ', '');
    const session = token
      ? await one('SELECT user_id FROM sessions WHERE token = $1 AND expires_at > now()', [hashSessionToken(token)])
      : null;
    if (!session) return reply.status(401).send({ error: 'Not authenticated' });
    const existing = await one('SELECT * FROM user_build_likes WHERE user_id = $1 AND build_id = $2', [session.user_id, id]);
    if (existing) {
      await one('DELETE FROM user_build_likes WHERE user_id = $1 AND build_id = $2', [session.user_id, id]);
      const row = await one('UPDATE builds SET likes = GREATEST(likes - 1, 0) WHERE id = $1 RETURNING likes', [id]);
      return { liked: false, likes: Number(row?.likes ?? 0) };
    } else {
      await one('INSERT INTO user_build_likes (user_id, build_id) VALUES ($1, $2)', [session.user_id, id]);
      const row = await one('UPDATE builds SET likes = likes + 1 WHERE id = $1 RETURNING likes', [id]);
      return { liked: true, likes: Number(row?.likes ?? 0) };
    }
  });
}
